src/ipred.rs

```diff
@@ -364,6 +364,29 @@ impl DcGen {
     }
 }
 
+/// Reconstructs the reference to the topleft edge array from a pointer into the
+/// array and an offset from the start of the array.
+///
+/// The topleft pointer passed to asm is always a pointer into a buffer of
+/// length [`SCRATCH_EDGE_LEN`]. For the Rust fallbacks we also pass in the
+/// offset from the front of the buffer so that we can reconstruct the original
+/// array reference in order to use safe array operations within the fallbacks.
+///
+/// # Safety
+///
+/// `topleft_ptr` must be a pointer into an array of length [`SCRATCH_EDGE_LEN`]
+/// and is `topleft_off` elements from the beginning of the array. This should
+/// be guaranteed by the logic in `angular_ipred::call`.
+unsafe fn reconstruct_topleft<'a, BD: BitDepth>(
```

Comment:
What does this lifetime do if it's not used in any argument?

Reply:
It's the lifetime of the returned reference. Since we're creating a reference from a raw pointer we need to explicitly annotate the output lifetime since there's no corresponding input lifetime.

Reply:
But we don't give it an explicit lifetime we can call it.  I'm not sure how this works exactly.

Reply:
It works the same way that dereferencing a raw pointer works. When you dereference a pointer Rust materializes an arbitrary lifetime based on how the returned reference is used. So here the output lifetime is effectively unconstrained, and will last as long as you use the reference since there's no source lifetime.